### ISSUE TITLE
Enhancement: Make unbound function pointers throw in debug mode

### ISSUE THREAD

**Issue by jpf91** [mdparker/Derelict3#85]
I hope this hasn't been proposed already, but have you considered setting unbound OpenGL functions to some error function? Basically:

``` d
template throwFunc(alias func)
{
    alias typeof(func) OriginalType;
    ReturnType!OriginalType throwFunc(ParameterTypeTuple!OriginalType)
    {
        throw new Exception("Function " ~ func.stringof ~ " is not bound!");
    }
}
```

It seems a little bit more difficult to implement for extern(C) functions, might have to use a string mixin:

``` d
void* throwFunc(alias func)()
{
    alias typeof(func) OriginalType;
    alias ReturnType!OriginalType RetType;
    alias ParameterTypeTuple!OriginalType ParamTypes;

    mixin("
    extern(C) " ~ RetType.stringof ~ " " ~ func.stringof ~"throwHelperDFKLS("~ParamTypes.stringof~")
    {
        throw new Exception(\"Function " ~ func.stringof ~ " is not bound!\");
    }
    return &"~ func.stringof~"throwHelperDFKLS;";
}

glGetError = throwFunc(glGetError);
```

it shouldn't run slower, but it might blow up executables a bit but in debug mode, that might be ok?

**Comment by mdparker** [comment]
Definitely something to think about.

**Comment by jpf91** [comment]
Here's some working code, you only have to decide which variant (if any) you want to use ;-)

http://dpaste.dzfl.pl/13e0da35

**Comment by mdparker** [comment]
I've looked at this now and again over the past few months and very nearly started to implement it at one point, but I hesitated. Having thought about it some more I'm not sure how practical it would be to add it to the binding. I'm not sure what the impact on compile time and bloat would be, but it would likely be noticeable. Is the added benefit worth it? Given that one can already prevent calling unloaded functions by checking for the currently loaded GLVersion and testing extensions before calling any of their functions, I don't believe so. Thanks for the idea and sample implementation.
